cmd/nsdns/watch.go

```diff
@@ -0,0 +1,194 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+import (
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+import (
+	"github.com/Eagerod/kube-namesilo-dns/pkg/icanhazip"
+	"github.com/Eagerod/kube-namesilo-dns/pkg/namesilo_api"
+	"github.com/Eagerod/kube-namesilo-dns/pkg/nsdns"
+)
+
+func watchCommand() *cobra.Command {
+	var ingressClass string
+	var domainName string
+
+	watchCmd := &cobra.Command{
+		Use:   "watch",
+		Short: "watch ingress objects, and create dns records",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			log.SetLevel(log.DebugLevel)
+
+			if domainName == "" {
+				return fmt.Errorf("Must provide a domain name to update DNS records.")
+			}
+
+			if ingressClass == "" {
+				return fmt.Errorf("Must provide an ingress class to select DNS records.")
+			}
+
+			nsApiKey := os.Getenv("NAMESILO_API_KEY")
+			if nsApiKey == "" {
+				return fmt.Errorf("Failed to find NAMESILO_API_KEY in environment. Cannot proceed.")
+			}
+
+			api := namesilo_api.NewNamesiloApi(domainName, nsApiKey)
+
+			rrMutex := sync.RWMutex{}
+			var records []namesilo_api.ResourceRecord
+			var ip string
+
+			refreshState := func() error {
+				var err error
+				rrMutex.Lock()
+				defer rrMutex.Unlock()
+
+				records, err = api.ListDNSRecords()
+				if err != nil {
+					return err
+				}
+				ip, err = icanhazip.GetPublicIP()
+				if err != nil {
+					return err
+				}
+
+				log.Debug("Updated local resource records and IP address")
+
+				return nil
+			}
+
+			if err := refreshState(); err != nil {
+				return err
+			}
+
+			// Don't actually start the informer until basic information is
+			//   available.
+			go func() {
+				for range time.Tick(time.Hour) {
```

Comment:
_Should_ probably be updated to be a `select` that exits the goroutine upon signal/context cancel kind of thing. Doesn't do anything other than refreshing local state though, so not really much harm in letting this exit "unsafely"